pysages/methods/ffs.py

```diff
@@ -0,0 +1,314 @@
+            write_to_file(K_t)
+
+
+def _ffs(method,snapshot,helpers):
+    cv = method.cv
+    dt = snapshot.dt
+    natoms = np.size(snapshot.positions, 0)
+
+    #initialize method
+    def initialize():
+        bias = np.zeros((natoms, 3))
+        xi, _ = cv(helpers.query(snapshot))
+        return FFSState(bias, xi)
+
+    def update(state, data):
+        xi, _ = cv(data)
+        bias = state.bias
+        return FFSState(bias, xi)
+
+    return snapshot, initialize, generalize(update,helpers)
+
+
+def write_to_file(value):
+    with open("ffs_results.dat", "a+") as f:
+        f.write(str(value) + "\n")
+
+
+# Since snapshots are depleted each window, this function restores the list to
+# its initial values. This only works with stochastic integrators like BD or
+# Langevin, for other, velocity resampling is needed
+def increase_snaps(windows, initial_w):
+    if len(windows) > 0:
+        ratio = len(initial_w) // len(windows)
+        windows = windows * ratio
+
+    return windows
+
+
+def check_input(grid, xi, verbose = False):
+    """Verify if the initial configuration is a good one."""
+    is_good = xi < grid[0]
+
+    if is_good:
+        print("Good initial configuration\n")
+        print(xi)
+    elif verbose:
+        print(xi)
+
+    return is_good
+
+
+def basin_sampling(
+    max_num_snapshots,
+    sampling_time,
+    grid,
+    run,
+    sampler,
+    reference_snapshot,
+    helpers,
+    cv
+):
+    """
+    Sampling of basing configurations for initial flux calculations.
+    """
+    basin_snapshots = []
+    win_A = grid[0]
+    xi = sampler.state.xi.block_until_ready()
+
+    print("Starting basin sampling\n")
+    while len(basin_snapshots) < int(max_num_snapshots):
+        run(sampling_time)
+        xi = sampler.state.xi.block_until_ready()
+
+        if np.all(xi < win_A):
+            snap = copy(sampler.snapshot)
+            basin_snapshots.append(snap)
+            print("Storing basing configuration with cv value:\n")
+            print(xi)
+        else:
+            helpers.restore(sampler.snapshot, reference_snapshot)
+            xi, _ = cv(helpers.query(sampler.snapshot))
+            print("Restoring basing configuration since system leave basin with cv value:\n")
+            print(xi)
+
+    print(f"Finish sampling Basin with {max_num_snapshots} snapshots\n")
+
+    return basin_snapshots
+
+
+def initial_flow(
+    Num_window0, timestep, grid, initial_snapshots, run, sampler, helpers, cv
+):
+    """Selects snapshots from list generated with `basin_sampling`."""
+
+    success = 0
+    time_count = 0.0
+    window0_snaps = []
+    win_A = grid[0]
+
+    for i in range(0, Num_window0):
+        print(f"Initial stored configuration: {i}\n")
+        helpers.restore(sampler.snapshot, initial_snapshots[i])
+        xi, _= cv(helpers.query(sampler.snapshot))
+        print(xi)
+
+        has_reached_A = False
+        while not has_reached_A:
+            # this can be used not every timestep
```

Comment:
It's not clear what is meant by this. Does it mean that we could change it so the argument to `run` is a different number of time steps?

Reply:
Yes, we can include an argument in run similar to sampling_steps_basin in order to monitor the CV not every timestep in this function.